portal/server/controllers/utilisation-report-service/previous-reports/helpers/previous-reports-view-model-mapper.test.ts

```diff
@@ -0,0 +1,250 @@
+import { UtilisationReportReconciliationStatus } from '@ukef/dtfs2-common';
+import { PRIMARY_NAV_KEY } from '../../../../constants';
+import { PreviousUtilisationReportsResponseBody, UtilisationReportResponseBody } from '../../../../api-response-types';
+import { mapToPreviousReportsViewModel } from './previous-reports-view-model-mapper';
+import { aPortalSessionUser } from '../../../../../test-helpers/test-data/portal-session-user';
+import { aUtilisationReportResponse } from '../../../../../test-helpers/test-data/utilisation-report';
+
+describe('previous-reports-view-model-mapper', () => {
+  describe('mapToPreviousReportsViewModel', () => {
+    it('returns with no navItems or reports and year undefined when there are no previous reports', () => {
+      // Arrange
+      const user = aPortalSessionUser();
+      const responseBody: PreviousUtilisationReportsResponseBody = [];
+
+      // Act
+      const result = mapToPreviousReportsViewModel('2023', user, responseBody);
+
+      // Assert
+      expect(result).toEqual({
+        user,
+        primaryNav: PRIMARY_NAV_KEY.PREVIOUS_REPORTS,
+        navItems: [],
+        reports: [],
+        year: undefined,
+      });
+    });
+
+    it('sets the year to the target year query when provided', () => {
+      // Arrange
+      const user = aPortalSessionUser();
+      const responseBody: PreviousUtilisationReportsResponseBody = [
+        { year: 2024, reports: [] },
+        { year: 2023, reports: [] },
+      ];
+      const targetYearQuery = '2023';
+
+      // Act
+      const result = mapToPreviousReportsViewModel(targetYearQuery, user, responseBody);
+
+      // Assert
+      expect(result.year).toEqual(2023);
+    });
+
+    it('sets the year to the year of the first response group if no target year query is undefined', () => {
+      // Arrange
+      const user = aPortalSessionUser();
+      const responseBody: PreviousUtilisationReportsResponseBody = [
+        { year: 2024, reports: [] },
+        { year: 2023, reports: [] },
+      ];
+      const targetYearQuery = undefined;
+
+      // Act
+      const result = mapToPreviousReportsViewModel(targetYearQuery, user, responseBody);
+
+      // Assert
+      expect(result.year).toEqual(2024);
+    });
+
+    it('creates a nav item for each year the response body contains a group of reports for with the target year active', () => {
+      // Arrange
+      const user = aPortalSessionUser();
+      const responseBody: PreviousUtilisationReportsResponseBody = [
+        { year: 2024, reports: [] },
+        { year: 2023, reports: [] },
+      ];
+
+      // Act
+      const result = mapToPreviousReportsViewModel('2023', user, responseBody);
+
+      // Assert
+      expect(result.navItems).toHaveLength(2);
+      expect(result.navItems[0]).toEqual({
```

Comment:
Does this work if you write `expect(result.navItems).toEqual([{ <navItem1> }, { <navItem2> }]);`?

Reply:
Yeah, I think to me it doesn't make much difference, generally I like try and keep my assertions as specific as possible (although this is covering quite a lot anyway I suppose actually). I can switch it if you think asserting the whole array at once is better?

Reply:
I think in this case asserting the whole array would allow you to capture both the length and item assertions in one line, but I don't particularly mind which one we go with

Reply:
I would say I don't think keeping assertions to as few lines as possible is important, in fact I would almost go the other way, in trying to have each way a test can break as it's own assertion so you immediately see which way the test broke when it breaks. It also gives you a good way of seeing if you are testing too much in one test, if you end up with lots of assertions you might actually be best off splitting the test rather than combining the assertions down into as few assertions as possible.

Reply:
I think jest is pretty good at showing you where the test has gone wrong even if you only do one object assertion, but I suppose it depends on how you read tests and what specifically goes wrong - I have seen cases where more assertions makes it harder to figure out where things went wrong and cases where fewer assertions do the same thing. I don't think it makes much difference either way in this particular case though and agree with the general principle that if you're testing a large object which has some mapped values, the tests themselves should probably be split up rather than just the assertions